File: book_catalogue/services/open_library/schemas/edition.py
from __future__ import annotations

__all__ = ["Edition"]

from datetime import date, datetime

from pydantic import Field

from book_catalogue.console import CONSOLE
from book_catalogue.services.open_library.schemas import (
    BaseModel,
    DatetimeResource,
    Link,
    Resource,
    TextResource,
)


class Contributor(BaseModel):
    name: str
    role: str


class Content(BaseModel):
    label: str | None = None
    level: int
    pagenum: str | None
    title: str
    type: Resource | None = None


class Identifiers(BaseModel):
    amazon: list[str] = Field(default_factory=list)
    better_world_books: list[str] = Field(default_factory=list)
    goodreads: list[str] = Field(default_factory=list)
    google: list[str] = Field(default_factory=list)
    librarything: list[str] = Field(default_factory=list)
    wikidata: list[str] = Field(default_factory=list)


class Edition(BaseModel):
    authors: list[Resource] = Field(default_factory=list)
    by_statement: str | None = None
    classifications: dict[int, int] = Field(default_factory=dict)
    contributions: list[str] = Field(default_factory=list)
    contributors: list[Contributor] = Field(default_factory=list)
    copyright_date: str | None = None
    covers: list[int] = Field(default_factory=list)
    created: DatetimeResource
    description: str | TextResource | None = None
    dewey_decimal_class: list[str] = Field(default_factory=list)
    edition_name: str | None = None
    first_sentence: str | TextResource | None = None
    full_title: str | None = None
    genres: list[str] = Field(default_factory=list)
    ia_box_id: list[str] = Field(default_factory=list)
    ia_loaded_id: list[str] = Field(default_factory=list)
    identifiers: Identifiers = Identifiers()
    isbn_10: list[str] = Field(default_factory=list)
    isbn_13: list[str] = Field(default_factory=list)
    key: str
    languages: list[Resource] = Field(default_factory=list)
    last_modified: DatetimeResource
    latest_revision: int
    lc_classifications: list[str] = Field(default_factory=list)
    lccn: list[str] = Field(default_factory=list)
    links: list[Link] = Field(default_factory=list)
    local_id: list[str] = Field(default_factory=list)
    location: list[str] = Field(default_factory=list)
    notes: str | TextResource | None = None
    number_of_pages: int | None = None
    ocaid: str | None = None
    oclc_numbers: list[str] = Field(default_factory=list)
    oclc_number: list[str] = Field(default_factory=list)
    other_titles: list[str] = Field(default_factory=list)
    pagination: str | None = None
    physical_dimensions: str | None = None
    physical_format: str | None = None
    publish_country: str | None = None
    publish_date: str | None = None
    publish_places: list[str] = Field(default_factory=list)
    publishers: list[str]
    revision: int
    series: list[str] = Field(default_factory=list)
    source_records: list[str] = Field(default_factory=list)
    subject_people: list[str] = Field(default_factory=list)
    subject_place: list[str] = Field(default_factory=list)
    subject_places: list[str] = Field(default_factory=list)
    subject_time: list[str] = Field(default_factory=list)
    subjects: list[str] = Field(default_factory=list)
    subtitle: str | None = None
    table_of_contents: list[Content] = Field(default_factory=list)
    title: str
    translated_from: list[Resource] = Field(default_factory=list)
    translation_of: str | None = None
    type: Resource
    weight: str | None = None
    work_title: list[str] = Field(default_factory=list)
    work_titles: list[str] = Field(default_factory=list)
    works: list[Resource]

    @property
    def edition_id(self) -> str:
        return self.key.split("/")[-1]

    def get_description(self) -> str | None:
        if self.description:
            if isinstance(self.description, TextResource):
                return self.description.value
            return self.description
        return None

    def get_publish_date(self) -> date | None:
        if not self.publish_date:
            return None
        try:
            return date.fromisoformat(self.publish_date)
        except ValueError as iso_err:
            for _format in [
                "%Y",
                "%b, %Y",
                "%Y-%b-%d",
                "%b %d, %Y",
                "%B %d, %Y",
                "%B %Y",
                "%d %b %Y",
            ]:
                try:
                    return datetime.strptime(self.publish_date, _format).date()
                except ValueError:
                    pass
            CONSOLE.print_exception(theme="ansi_dark")
            raise iso_err

File: openlibrary/src/main/kotlin/github/buriedincode/openlibrary/schemas/Edition.kt
package github.buriedincode.openlibrary.schemas

import github.buriedincode.openlibrary.serializers.DescriptionSerializer
import github.buriedincode.openlibrary.serializers.LocalDateSerializer
import github.buriedincode.openlibrary.serializers.LocalDateTimeSerializer
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonNames

@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class Edition(
    val authors: List<Resource> = emptyList(),
    val byStatement: String? = null,
    val classifications: Classification? = null,
    val contributions: List<String> = emptyList(),
    val contributors: List<Contributor> = emptyList(),
    @Serializable(with = LocalDateSerializer::class)
    val copyrightDate: LocalDate? = null,
    val covers: List<Int> = emptyList(),
    @Serializable(with = LocalDateTimeSerializer::class)
    val created: LocalDateTime? = null,
    @Serializable(with = DescriptionSerializer::class)
    val description: String? = null,
    val deweyDecimalClass: List<String> = emptyList(),
    val editionName: String? = null,
    @Serializable(with = DescriptionSerializer::class)
    val firstSentence: String? = null,
    val fullTitle: String? = null,
    val genres: List<String> = emptyList(),
    val iaBoxId: List<String> = emptyList(),
    val iaLoadedId: List<String> = emptyList(),
    val identifiers: Identifiers? = null,
    @JsonNames("isbn_10")
    val isbn10: List<String> = emptyList(),
    @JsonNames("isbn_13")
    val isbn13: List<String> = emptyList(),
    val key: String,
    val languages: List<Resource> = emptyList(),
    @Serializable(with = LocalDateTimeSerializer::class)
    val lastModified: LocalDateTime? = null,
    val latestRevision: Int,
    val lcClassifications: List<String> = emptyList(),
    val lccn: List<String> = emptyList(),
    val links: List<Link> = emptyList(),
    val localId: List<String> = emptyList(),
    val location: List<String> = emptyList(),
    @Serializable(with = DescriptionSerializer::class)
    val notes: String? = null,
    val numberOfPages: Int? = null,
    val ocaid: String? = null,
    val oclcNumber: List<String> = emptyList(),
    val oclcNumbers: List<String> = emptyList(),
    val otherTitles: List<String> = emptyList(),
    val pagination: String? = null,
    val physicalDimensions: String? = null,
    val physicalFormat: String? = null,
    val publishCountry: String? = null,
    @Serializable(with = LocalDateSerializer::class)
    val publishDate: LocalDate? = null,
    val publishPlaces: List<String> = emptyList(),
    val publishers: List<String> = emptyList(),
    val revision: Int,
    val series: List<String> = emptyList(),
    val sourceRecords: List<String> = emptyList(),
    val subjectPeople: List<String> = emptyList(),
    val subjectPlace: List<String> = emptyList(),
    val subjectPlaces: List<String> = emptyList(),
    val subjects: List<String> = emptyList(),
    val subjectTime: List<String> = emptyList(),
    val subtitle: String? = null,
    val tableOfContents: List<Content> = emptyList(),
    val translatedFrom: List<Resource> = emptyList(),
    val translationOf: String? = null,
    val title: String,
    val type: Resource,
    val uriDescriptions: List<String> = emptyList(),
    val uris: List<String> = emptyList(),
    val url: List<String> = emptyList(),
    val weight: String? = null,
    val works: List<Resource> = emptyList(),
    val workTitle: List<String> = emptyList(),
    val workTitles: List<String> = emptyList(),
) {
    @Serializable
    data class Classification(
        val key: String? = null,
    )

    @Serializable
    data class Content(
        val label: String? = null,
        val level: Int,
        val pagenum: String? = null,
        val title: String,
        val type: Resource? = null,
    )

    @Serializable
    data class Contributor(
        val name: String,
        val role: String,
    )

    @OptIn(ExperimentalSerializationApi::class)
    @Serializable
    data class Identifiers(
        val amazon: List<String> = emptyList(),
        @JsonNames("amazon.co.uk_asin")
        val amazonCoUkAsin: List<String> = emptyList(),
        val betterWorldBooks: List<String> = emptyList(),
        val goodreads: List<String> = emptyList(),
        val google: List<String> = emptyList(),
        val issn: List<String> = emptyList(),
        val librarything: List<String> = emptyList(),
        val wikidata: List<String> = emptyList(),
    )
}
